From 95eebd85974b108d7bee4490962b585170f0086b Mon Sep 17 00:00:00 2001
From: Gustavo Duarte <gduarte@redhat.com>
Date: Wed, 3 Dec 2014 17:21:46 -0500
Subject: [char] tpm: Fix NULL return in tpm_ibmvtpm_get_desired_dma

Message-id: <1417627306-23002-1-git-send-email-gduarte@redhat.com>
Patchwork-id: 101114
O-Subject: [RHEL7.1 PATCH BZ 1154818] tpm: Fix NULL return in tpm_ibmvtpm_get_desired_dma
Bugzilla: 1154818
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Don Zickus <dzickus@redhat.com>

BZ 1154818
Upstream Status: https://github.com/PeterHuewe/linux-tpmdd/commit/7a4353be2d2467f73a5806904447cb2d977f26a0
Build Info: https://brewweb.devel.redhat.com/taskinfo?taskID=8324864
ppc64le: https://brewweb.devel.redhat.com/taskinfo?taskID=8324872
Tested: Kernel boots fine on ppc64 and ppc64le.

Author: "Hon Ching (Vicky) Lo" <honclo@linux.vnet.ibm.com>
Date:   Sun, 30 Nov 2014 15:01:28 +0100

    tpm: Fix NULL return in tpm_ibmvtpm_get_desired_dma

    There was an oops in tpm_ibmvtpm_get_desired_dma, which caused
    kernel panic during boot when vTPM is enabled in Power partition
    configured in AMS mode.

    vio_bus_probe calls vio_cmo_bus_probe which calls
    tpm_ibmvtpm_get_desired_dma to get the size needed for DMA allocation.
    The problem is, vio_cmo_bus_probe is called before calling probe, which
    for vtpm is tpm_ibmvtpm_probe and it's this function that initializes
    and sets up vtpm's CRQ and gets required data values.  Therefore,
    since this has not yet been done, NULL is returned in attempt to get
    the size for DMA allocation.

    We added a NULL check.  In addition, a default buffer size will
    be set when NULL is returned.

    Cc: <stable@vger.kernel.org>
    Signed-off-by: Hon Ching (Vicky) Lo <honclo@linux.vnet.ibm.com>
    Signed-off-by: Peter Huewe <peterhuewe@gmx.de>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/char/tpm/tpm_ibmvtpm.c b/drivers/char/tpm/tpm_ibmvtpm.c
index 56b07c3..010d814 100644
--- a/drivers/char/tpm/tpm_ibmvtpm.c
+++ b/drivers/char/tpm/tpm_ibmvtpm.c
@@ -307,6 +307,14 @@ static int tpm_ibmvtpm_remove(struct vio_dev *vdev)
 static unsigned long tpm_ibmvtpm_get_desired_dma(struct vio_dev *vdev)
 {
  struct ibmvtpm_dev *ibmvtpm = ibmvtpm_get_data(&vdev->dev);
+
+ /* ibmvtpm initializes at probe time, so the data we are
+ * asking for may not be set yet. Estimate that 4K required
+ * for TCE-mapped buffer in addition to CRQ.
+ */
+ if (!ibmvtpm)
+  return CRQ_RES_BUF_SIZE + PAGE_SIZE;
+
  return CRQ_RES_BUF_SIZE + ibmvtpm->rtce_size;
 }
 
-- 
1.7.1